parsers/CZ.py

```diff
@@ -0,0 +1,221 @@
+    "HPP": "hydro",  # Water
+    "PsPP": "hydro",  # Pump Water storage
+    "AltPP": "biomass",  # Alternative
+    "ApPP": "unknown",  # factory
+    "PVPP": "solar",  # photovoltaic
+    "WPP": "wind",  # wind
+    "unknown": "unknown",
+}
+translate_table_dist = {
+    "SEPS": "SVK",
+    "APG": "AT",
+    "PSE": "PL",
+    "TenneT": "DE",
+    "50HzT": "DE",
+}
+url = "https://wwwtest.ceps.cz/_layouts/CepsData.asmx"
+
+
+def get_mapper(xmlload):
+    series = xmlload.find("series")
+    mapping = {}
+    for tag in series:
+        generator = tag["name"].replace(" [MW]", "")
+        mapping[generator] = tag["id"]
+
+    return mapping
+
+
+def make_request(session, payload, zone_key):
+    headers = {
+        "Content-Type": "application/soap+xml; charset=utf-8",
+        "Content-Length": "1",
+    }
+
+    r = session or Session()
+    res = r.post(url, headers=headers, data=payload)
+    assert res.status_code == 200, (
+        "Exception when fetching production for "
+        "{}: error when calling url={}".format(zone_key, url)
+    )
+
+    return res
+
+
+def fetch_production(
+    zone_key: str = "CZ",
+    session: Optional[Session] = None,
+    target_datetime: Optional[datetime] = None,
+    logger: Logger = getLogger(__name__),
+) -> Union[List[dict], dict]:
+
+    if not target_datetime:
+        target_datetime = datetime.now().replace(minute=0, second=0, microsecond=0)
+    from_datetime = target_datetime - timedelta(hours=1)
+
+    payload = """<?xml version="1.0" encoding="utf-8"?>
+            <soap12:Envelope xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:soap12="http://www.w3.org/2003/05/soap-envelope">
+              <soap12:Body>
+                <Generation xmlns="https://www.ceps.cz/CepsData/">
+                  <dateFrom>{0}</dateFrom>
+                  <dateTo>{1}</dateTo>
+                  <agregation>{2}</agregation>
+                  <function>{3}</function>
+                  <version>{4}</version>
+                  <para1>{5}</para1>
+                </Generation>
+              </soap12:Body>
+            </soap12:Envelope>""".format(
+        from_datetime.isoformat(), target_datetime.isoformat(), "QH", "AVG", "RT", "all"
+    )
+
+    content = make_request(session, payload, zone_key).content
+    xml = BeautifulSoup(content, "xml")
+    mapper = get_mapper(xml)
+
+    data_tag = xml.find("data")
+    data_list = []
+
+    for values in data_tag:
+        data = {
+            "zoneKey": zone_key,
+            "production": {},
+            "storage": {},
+            "source": url,
+            "datetime": datetime.fromisoformat(values["date"]).isoformat(),
+        }
+
+        for k, v in mapper.items():
+            generator = translate_table_gen[k]
+            if k != "PsPP":
+                data["production"][generator] = float(values[v])
+            else:
+                data["storage"][generator] = float(values[v])
```

Comment:
```suggestion
            else:
                data["storage"][generator] = float(values[v]) * -1
```
It's a bit counter intuetive but we currently treat negative values as production and positives ones as "charging"

Reply:
this will be problematic. CEPS is not publishing data about charging. I can write an email to them asking whether it is somewhere in the data set but I couldn't find it. It simply returns positive values once producing nothing more.

Reply:
That's fine (I think) but positive production need to be returned as a negative value.

(We really want to change this but fixing this tiny thing would require a lot of changes 😅)